### PR TITLE
[NewPM/CodeGen] Rewrite pass manager nesting

### PR DESCRIPTION
**Pull request by aeubanks**
Currently the new PM infra for codegen puts everything into a MachineFunctionPassManager. The MachineFunctionPassManager owns both Module passes and MachineFunction passes, and batches adjacent MachineFunction passes like a typical PassManager.

The current MachineFunctionAnalysisManager also directly references a module and function analysis manager to get results.

The initial argument was that the codegen pipeline is relatively "flat", meaning it's mostly machine function passes with a couple of module passes here and there. However, there are a couple of issues with this as compared to a more structured nesting more like the optimization pipeline. For example, it doesn't allow running function passes then machine function passes on a function and its machine function all at once. It also currently requires the caller to split out the IR passes into one pass manager and the MIR passes into another pass manager.

This patch rewrites the new pass manager infra for the codegen pipeline to be more similar to the nesting in the optimization pipeline. Basically, a Function contains a MachineFunction. So we can have Module -> Function -> MachineFunction adaptors. It also rewrites the analysis managers to have inner/outer proxies like the ones in the optimization pipeline. The new pass managers/adaptors/analysis managers can be seen in use in PassManagerTest.cpp.

This allows us to consolidate to just having to add to one ModulePassManager when using the codegen pipeline.

I haven't added the Function -> MachineFunction adaptor in this patch, but it should be added when we merge AddIRPass/AddMachinePass so that we can run IR and MIR passes on a function before proceeding to the next function.

The MachineFunctionProperties infra for MIR verification is still WIP.

### CODE REVIEW

**Comment by aeubanks** [comment]
as mentioned in the description the MachineFunctionProperties  stuff is still WIP, but I'd like some eyes on this now

we should extend https://llvm.org/docs/NewPassManager.html to talk about the design of the codegen infra

**Review by paperchalice — COMMENTED** [review]
Thanks for doing this! Some comments.

File: llvm/include/llvm/CodeGen/MachinePassManager.h [hunk 1]
```diff
@@ -46,217 +47,247 @@ struct MachinePassInfoMixin : public PassInfoMixin<DerivedT> {
...
+    InnerAnalysisManagerProxy<MachineFunctionAnalysisManager, Module>;
+
+template <>
+bool MachineFunctionAnalysisManagerModuleProxy::Result::invalidate(
+    Module &M, const PreservedAnalyses &PA,
+    ModuleAnalysisManager::Invalidator &Inv);
+extern template class InnerAnalysisManagerProxy<MachineFunctionAnalysisManager,
+                                                Module>;
+
+extern template class OuterAnalysisManagerProxy<ModuleAnalysisManager,
```
[the remark this answers is not in the document]
Comment (aeubanks) [reply llvm/include/llvm/CodeGen/MachinePassManager.h:173]: The parts about accessing outer analyses in https://llvm.org/docs/NewPassManager.html#using-analyses is still relevant for MachineFunction passes and Module analyses, so we should have these.

Agreed that MachineFunction passes should be able to query for Function analyses though.

**Review by paperchalice — CHANGES_REQUESTED** [review]

File: llvm/tools/llc/NewPMDriver.cpp [hunk 2]
```diff
@@ -182,49 +147,46 @@ int llvm::compileModuleWithNewPM(
...
+  if (Context.getDiagHandlerPtr()->HasErrors)
+    exit(1);
+
+  // Before executing passes, print the final values of the LLVM options.
+  cl::PrintOptionValues();
+
+  MPM.run(*M, MAM);
+
+  if (Context.getDiagHandlerPtr()->HasErrors)
+    exit(1);
```
[the remark this answers is not in the document]
Comment (aeubanks) [reply llvm/tools/llc/NewPMDriver.cpp:189]: oops, bad merge

File: llvm/tools/llc/NewPMDriver.cpp [hunk 3]
```diff
@@ -182,49 +147,46 @@ int llvm::compileModuleWithNewPM(
       return 1;
     }
 
-    ExitOnErr(PB.parsePassPipeline(MFPM, PassPipeline));
+    ExitOnErr(PB.parsePassPipeline(MPM, PassPipeline));
```
[the remark this answers is not in the document]
Comment (aeubanks) [reply llvm/tools/llc/NewPMDriver.cpp:150]: yes, although I'll leave that for a follow-up patch. added a FIXME

**Review by paperchalice — APPROVED** [review]

**Review by alinas — APPROVED** [review]
Thank you for working on this.

**Pull request merged**